helix-view/src/view.rs

```diff
@@ -63,13 +67,23 @@ impl JumpList {
 
     // Taking view and doc to prevent unnecessary cloning when jump is not required.
     pub fn backward(&mut self, view_id: ViewId, doc: &mut Document, count: usize) -> Option<&Jump> {
-        if let Some(current) = self.current.checked_sub(count) {
+        if let Some(mut current) = self.current.checked_sub(count) {
+            let cur_jump = (doc.id(), doc.selection(view_id).clone());
             if self.current == self.jumps.len() {
-                let jump = (doc.id(), doc.selection(view_id).clone());
-                self.push(jump);
+                let num_removed = self.push(cur_jump.clone());
+                current = current.saturating_sub(num_removed);
             }
             self.current = current;
-            self.jumps.get(self.current)
+            self.jumps.get(self.current).and_then(|prev_jump| {
+                // If a jumplist position is saved and then `jump_backward` is immediately invoked
+                // without moving the cursor, jump back one more step to prevent a no-op
+                if prev_jump == &cur_jump {
+                    self.current = self.current.saturating_sub(1);
+                    self.jumps.get(self.current)
+                } else {
+                    Some(prev_jump)
+                }
+            })
```

Comment:
Changing the behavior here sounds good to me but the `and_then` can be a little tough to read. Let's use the `?` operator to clean it up and also avoid making another clone of the selection to check equality (i.e. not binding `cur_jump` above):

```rust
// Avoid jumping to the current location.
let jump @ (doc_id, selection) = self.jumps.get(self.current)?;
if doc.id() == *doc_id && doc.selection(view_id) == selection {
    self.current = self.current.checked_sub(1)?;
    self.jumps.get(self.current)
} else {
    Some(jump)
}
```

Reply:
Looks good - updated